omp/matrix/csr_kernels.cpp

```diff
@@ -742,6 +744,49 @@ GKO_INSTANTIATE_FOR_EACH_VALUE_AND_INDEX_TYPE(
     GKO_DECLARE_CSR_CALC_NNZ_PER_ROW_IN_SPAN_KERNEL);
 
 
+template <typename ValueType, typename IndexType>
+void calculate_nonzeros_per_row_in_index_set(
+    std::shared_ptr<const DefaultExecutor> exec,
+    const matrix::Csr<ValueType, IndexType>* source,
+    const IndexSet<IndexType>& row_index_set,
+    const IndexSet<IndexType>& col_index_set, Array<IndexType>* row_nnz)
+{
+    auto num_row_subsets = row_index_set.get_num_subsets();
+    auto row_subset_begin = row_index_set.get_subsets_begin();
+    auto row_subset_end = row_index_set.get_subsets_end();
+    auto src_ptrs = source->get_const_row_ptrs();
+#pragma omp parallel for
+    for (size_type set = 0; set < num_row_subsets; ++set) {
+        for (size_type row = row_subset_begin[set]; row < row_subset_end[set];
+             ++row) {
+            row_nnz->get_data()[row - row_subset_begin[set]] =
+                zero<IndexType>();
+            Array<IndexType> l_idxs(
+                exec,
+                static_cast<size_type>(src_ptrs[row + 1] - src_ptrs[row]));
+            gko::kernels::omp::index_set::global_to_local(
+                exec, col_index_set.get_size(), col_index_set.get_num_subsets(),
+                col_index_set.get_subsets_begin(),
+                col_index_set.get_subsets_end(),
+                col_index_set.get_superset_indices(),
+                static_cast<IndexType>(l_idxs.get_num_elems()),
+                source->get_const_col_idxs() + src_ptrs[row], l_idxs.get_data(),
+                false);
```

Comment:
This seems really inefficient to me - it needs a lot of space even if extracting a small submatrix, and the information isn't reused between the two kernel calls. If we force the index set to be sorted (which IMO is a good invariant to have independent of this kernel), we can use upper_bound and cache the previous result, like we do in some Partition kernels.

Reply:
I am not sure you can re-use the information between two calls. The sets of global indices can be completely different. Unless you cache the local indices of the entire col_idxs of the parent matrix, I think it might be better to look-up the indices in every row that is relevant to the submatrix ? 

And the index_set itself is always sorted, but I am not sure we can guarantee that the indices themselves that we need to convert are sorted.

Reply:
Ah sorry, I mixed up the loops. I thought you were computing the local indices for the entire slice. Then my comment is more specific: Allocating inside such a hot loop will not give you good performance, and you read each output entry of the global_to_local kernel only once. A solution without allocation would be to use the inner kernel from global_to_local directly.

Reply:
You could also allocate an array with the maximal number of nnz per row outside these loops and then use a view of that, if you want to keep the kernel call.

Reply:
Yes, that is what I went with now for the reference version.

Reply:
It only computes the number of valid elements, so it shoule not need to convert global to local with allocation.
going through the intexset and the col index is enough to get the number.

Reply:
you can also use the submatrix row pointer information to create the compete `l_idx` out of the for-loop in `compute_submatrix_from_index_set` if you do not want to compute the maximum per each set.

Reply:
Going through the index_set/querying if an index is in the index set requires conversion from global_to_local. I moved that out of the loop with max_nnz_per_row for both ref and omp now. 

Yes, I can use to compute the entire l_idx outside the for loop, but I feel that is unnecessary and requires more memory ?

Reply:
the first part of global_to_local without allocation checks the validation and it does not transfer the data to local yet
```
for (size_type i = 0; i < num_indices; ++i) {
    auto index = global_indices[i];
    if (index >= index_space_size) {
        continue;
    }
    const auto bucket = std::distance(
        subset_begin,
        std::upper_bound(subset_begin, subset_begin + num_subsets, index));
    auto shifted_bucket = bucket == 0 ? 0 : (bucket - 1);
    if (subset_end[shifted_bucket] <= index) {
        continue;
    }
    nnz[res_row]++;
}
```